### PR TITLE
chore: separate server for each service

### LINKED ISSUE #74: Go SDK Changes

### PR DESCRIPTION
**Pull request by yhl25**
- separate servers for each operation(`map`, `reduce`, `mapstream`, `sink` and `sourcetransformer`)
- introduced `Server` interface inside `numaflow` package, all the servers will implement this interface
- Separate proto files for each operation and created separate types for each server to avoid confusion
- Separate sock address for each operation
- refactored directories to avoid confusion between the operations 
- refactored the interfaces exposed to users to remove handlers
- created separate types `Message`, `MessageBuilder` etc for each service.

fixes #74

### CODE REVIEW

File: pkg/apis/proto/function/reducefn/reduce.proto [hunk 1]
```diff
@@ -0,0 +1,62 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/function/reducefn";
```
Comment (whynowy) [inline pkg/apis/proto/function/reducefn/reduce.proto:3]: A generic question, any reason of adding a tailing `-fn` to the package?

Comment (yhl25) [reply pkg/apis/proto/function/reducefn/reduce.proto:3]: map, reduce package names are already present, so added fn here. I couldn't come up with a better package name :)

**Comment by vigith** [comment]
their proto files are complex; hence they need it? do we really have a use-case where we will use versioning?

**Comment by whynowy** [comment]
> their proto files are complex; hence they need it? do we really have a use-case where we will use versioning?

Complexity shouldn't be a factor of doing versioning or not. Versioning give us a chance to run backward compatible code. Assume some day, there's a breaking change on the contract (v2), the platform code may have a chance to know which version of code the ud container is running, and use corresponding version of client to do the communication.

File: pkg/apis/proto/reduce/v1/reduce.proto [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+message ReduceResponse {
+  repeated string keys = 1;
+  bytes value = 2;
+  repeated string tags = 3;
+}
+
+/**
+ * ReduceResponseList represents a list of response elements.
+ */
+message ReduceResponseList {
```
Comment (vigith) [inline pkg/apis/proto/reduce/v1/reduce.proto:41]: remove List

File: pkg/apis/proto/sourcetransformer/v1/transformer.proto [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  repeated string keys = 1;
+  bytes value = 2;
+  EventTime event_time = 3;
+  repeated string tags = 4;
+}
+
+/**
+ * SourceTransformerResponseList represents a list of response elements.
+ */
+message SourceTransformerResponseList {
```
Comment (vigith) [inline pkg/apis/proto/sourcetransformer/v1/transformer.proto:56]: remove list

File: pkg/apis/proto/sourcetransformer/v1/transformer.proto [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+message EventTime {
+  // event_time is the time associated with each datum.
+  google.protobuf.Timestamp event_time = 1;
+  // future we can add an enum to represent whether the event time is
+  // based on INGEST_TIME or EVENT_TIME or PROCESSING_TIME
+}
+
+message Watermark {
+  // watermark is the monotonically increasing time which denotes completeness for the given time for the given vertex.
+  google.protobuf.Timestamp watermark = 1;
```
Comment (vigith) [inline pkg/apis/proto/sourcetransformer/v1/transformer.proto:29]: remove this indirection?

File: pkg/map/doc.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+	func main() {
+		err := _map.NewServer(_map.MapperFunc(Map)).Start(context.Background())
+		if err != nil {
+			log.Panic("Failed to start map function server: ", err)
+		}
+	}
+*/
+
+package _map
```
Comment (vigith) [inline pkg/map/doc.go:35]:
```suggestion
package map_
```

not natural for users to start typing with `_`, the prefix underscore will be hinted by the IDE.

**Review by vigith — COMMENTED** [review]
partial review

File: pkg/sourcetransformer/interface.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+type Datum interface {
+	Value() []byte
+	EventTime() time.Time
+	Watermark() time.Time
+}
+
+// SourceTransformer is the interface of SourceTransformer function implementation.
+type SourceTransformer interface {
+	// SourceTransformer is the function to process each coming message.
+	SourceTransformer(ctx context.Context, keys []string, datum Datum) Messages
```
Comment (KeranYang) [inline pkg/sourcetransformer/interface.go:18]: nit: use `Transform` as the name. It's a verb and a bit shorter.

File: pkg/apis/proto/mapstream/v1/mapstream.proto [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+
+
+package mapstream.v1;
+
+service MapStream {
+  // MapStreamFn applies a function to each request element and returns a stream.
+  rpc MapStreamFn(MapStreamRequest) returns (stream MapStreamResponse.Result);
```
Comment (KeranYang) [inline pkg/apis/proto/mapstream/v1/mapstream.proto:13]:
```suggestion
  rpc MapStreamFn(MapStreamRequest) returns (stream MapStreamResponse);
```

File: pkg/apis/proto/mapstream/v1/mapstream.proto [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * MapStreamResponse represents a response element.
+ */
+message MapStreamResponse {
+  message Result {
+    repeated string keys = 1;
+    bytes value = 2;
+    repeated string tags = 3;
+  }
+  repeated Result results = 1;
```
Comment (KeranYang) [inline pkg/apis/proto/mapstream/v1/mapstream.proto:38]:
```suggestion
  Result result = 1;
```

File: pkg/apis/proto/mapstream/v1/mapstream.proto [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  rpc MapStreamFn(MapStreamRequest) returns (stream MapStreamResponse.Result);
+
+  // IsReady is the heartbeat endpoint for gRPC.
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
+}
+
+/**
+ * MapStreamRequest represents a request element.
+ */
+message MapStreamRequest {
```
Comment (KeranYang) [inline pkg/apis/proto/mapstream/v1/mapstream.proto:22]: Could you also make Request nested? Same for other protos

File: pkg/apis/proto/mapstream/v1/mapstream.proto [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  bytes value = 2;
+  google.protobuf.Timestamp event_time = 3;
+  google.protobuf.Timestamp watermark = 4;
+}
+
+/**
+ * MapStreamResponse represents a response element.
+ */
+message MapStreamResponse {
+  message Result {
```
Comment (KeranYang) [inline pkg/apis/proto/mapstream/v1/mapstream.proto:33]: Do we remove the event time and watermark from the response?

File: pkg/apis/proto/reduce/v1/reduce.proto [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  bytes value = 2;
+  google.protobuf.Timestamp event_time = 3;
+  google.protobuf.Timestamp watermark = 4;
+}
+
+/**
+ * ReduceResponse represents a response element.
+ */
+message ReduceResponse {
+  message Result {
```
Comment (KeranYang) [inline pkg/apis/proto/reduce/v1/reduce.proto:33]: Do we remove the event time and watermark from the response?

File: pkg/apis/proto/map/v1/map.proto [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  bytes value = 2;
+  google.protobuf.Timestamp event_time = 3;
+  google.protobuf.Timestamp watermark = 4;
+}
+
+/**
+ * MapResponse represents a response element.
+ */
+message MapResponse {
+  message Result {
```
Comment (KeranYang) [inline pkg/apis/proto/map/v1/map.proto:32]: Do we remove the event time and watermark from the response?

File: pkg/apis/proto/mapstream/v1/mapstream.proto (hunk 10 above)
Comment (yhl25) [reply pkg/apis/proto/mapstream/v1/mapstream.proto:33]: yes

File: pkg/apis/proto/mapstream/v1/mapstream.proto (hunk 7 above)
Comment (yhl25) [reply pkg/apis/proto/mapstream/v1/mapstream.proto]: done

**Review by KeranYang — APPROVED** [review]

File: pkg/apis/proto/sourcetransformer/v1/transformer.proto [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package sourcetransformer.v1;
+
+service SourceTransformer {
+  // SourceTransformFn applies a function to each request element.
+  // In addition to map function, SourceTransformFn also supports assigning a new event time to response.
+  // SourceTransformFn can be used only at source vertex by source data transformer.
+  rpc SourceTransformFn(SourceTransformerRequest) returns (SourceTransformerResponse);
```
Comment (KeranYang) [inline pkg/apis/proto/sourcetransformer/v1/transformer.proto:14]:
```suggestion
  rpc SourceTransformFn(SourceTransformRequest) returns (SourceTransformResponse);
```

File: pkg/apis/proto/sink/v1/sink.proto [hunk 14]
```diff
@@ -57,9 +47,22 @@ message Response {
...
+ * SinkResponseList is the list of responses. The number of elements in this list will be equal to the number of requests
+ * passed to the SinkFn.
  */
-message ResponseList {
-  repeated Response responses = 1;
+message SinkResponseList {
+  repeated SinkResponse responses = 1;
+}
+
+message EventTime {
```
[the remark this answers is not in the document]
Comment (KeranYang) [reply pkg/apis/proto/sink/v1/sink.proto:52]: Please address this comment :)

**Comment by vigith** [comment]
@yhl25, please do the non-critical asks in follow-up PR, i am going to merge this PR because it is a huge PR, and all major comments have been addressed.

**Pull request merged**